17 divisions within a list

tycebrown opened the pull request:
_Dependent on issue #16 (edit a list's title), pull request #23_ 

This one I jumped into too hastily. I should have first planned and discussed the design with you.

I made some major changes here. You might have to clear the database on your device before this works.

Points/Considerations/Questions (it's a lot):

- I went with an approach where each `ListModel`, instead of storing links to individual items, stores links to item _groups_ (which each store links to their items). This had some major impact on the way items are accessed, moved, etc. What would be a better approach? How should this be changed? 
- Each item stores an [isar backlink](https://isar.dev/links.html#backlinks) pointing to its group. From a test I ran (do you want me to share it?), it seems this backlink stores a _copy_ of its value (what it is linked to). This means that every item stores a copy of its item group. This copy of the item group, along with the corresponding item group in the list model, has to be refreshed often. Do you want me to refactor my design? Perhaps, instead of storing a backlink, simply store an id to an item group? I think this would simplify my code, as well as satisfy the efficiency freak in me. What do you think?
- I've heard of the [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter). Should this be followed? If not, what "law" should I go with? In either case, how should my code be refactored?
- Instead of simply having a single item group class, I created three classes: 
  - an abstract base class (`ItemGroup`), which provides an abstract `title` getter and setter, and a way of getting an `Iterable` of the underlying items of this item group.
  - the `ListModelItemGroup` (a concrete class that can be stored in a database and is used by `ListModel`), which extends `ItemGroup`. 
  - the `ItemGroupSearchResults`, which also extends `ItemGroup`. This class contains a wrapper of another `ItemGroup` along with an `Iterable<Item>`, which contain the results of a search within that group. Returned by `ItemGroup.searchItems()`.

  This was my first approach to the problem of being able, in a `ListWidget`, to display both item groups straight from a `ListModel` and from the results of a search query in a list. How should this be changed? What's a better solution; for example, having an item group class (maybe just called `ItemGroup`) and a `DisplayableItemGroup` (or something) that is used in a `ListWidget`. 
- One workaround I did which I think is clumsy (but am not sure how to change) is that in `ItemGroup`, I added a function called `asListModelItemGroup()` 
- The UI is a bit lazy, especially the group-picker for an item and the add-a-group button. How should this be changed?
- What do I need to rename?
- Anything else? Any tips, general advice, or design patterns?

tycebrown committed:
Merge branch '9-multiple-lists' into 17-divisions-within-a-list

tycebrown committed:
Merge branch '16-edit-lists-title' into 17-divisions-within-a-list

tycebrown committed:
unclean, ugly, code. set up for functionality

tycebrown committed:
Merge branch '10-settings' into 17-divisions-within-a-list

tycebrown committed:
Merge branch '13-better-editing' into 17-divisions-within-a-list

tycebrown committed:
Merge branch '16-edit-lists-title' into 17-divisions-within-a-list

tycebrown committed:
Raw,filthy, largely untested code for storing groups, displaying groups, but not adding groups

tycebrown committed:
Previous functionality now works with groups. Needs serious cleaning and way to add more groups.

tycebrown committed:
Functionality to add groups and basic title editing

tycebrown committed:
Now, an ItemGroup's title is the default value in an EditItemGroupDialog

tycebrown committed:
Now, list items can be moved between groups and groups added

tycebrown committed:
Code sweeping.

tycebrown committed:
Created the ItemGroupWidget, slight refactoring

tycebrown committed:
moved move from Item to ListModel

Commit:
Merge branch 'main' into 17-divisions-within-a-list

tycebrown committed:
Merge branch '15-search-in-single-list' into 17-divisions-within-a-list

Commit:
Merge branch 'main' into 17-divisions-within-a-list

Pertempto commented:
@tycebrown I can review this now? Since #23 is merged

tycebrown commented:
> @tycebrown I can review this now? Since #23 is merged

Yeah, though, like said, you might have to clear the database before it runs properly.

Will look into [isar data migration](https://isar.dev/recipes/data_migration.html).

tycebrown commented:
> Will look into [isar data migration](https://isar.dev/recipes/data_migration.html).

After a serious look and a failed attempt, I think (unless otherwise instructed) that I am going to hold off trying to code for data migration.

In fact, is now even the time to worry about that?

Pertempto commented:
> > Will look into [isar data migration](https://isar.dev/recipes/data_migration.html).
> 
> After a serious look and a failed attempt, I think (unless otherwise instructed) that I am going to hold off trying to code for data migration.
> 
> In fact, is now even the time to worry about that?

At this point we have no "real" users (the app is in what would be called the "alpha" stage), so it is totally fine to make breaking changes like completely changing the DB structure. Once an app is in later stages with many users, that is when you need to be careful not to introduce breaking changes.

Pertempto commented:
Can you please resolve merge conflicts before I review

tycebrown commented:
> Can you please resolve merge conflicts before I review

Oh, yes, I'm sorry, will do.

tycebrown committed:
Merge branch '25-make-checklist-creation-easier' into 17-divisions-within-a-list

tycebrown committed:
finished up merge

tycebrown committed:
Merge branch 'main' into 17-divisions-within-a-list

Pertempto commented on lib/model/item_group_search_results.dart:
```diff
@@ -0,0 +1,30 @@
+
+/// ItemGroupSearchResults:
+///   - an `ItemGroup` that contains the results of searching in another `ItemGroup`
+class ItemGroupSearchResults extends ItemGroup {
```
```suggestion
class FilteredItemGroup extends AbstractItemGroup {
```

Pertempto commented on lib/model/item_group.dart:
```diff
@@ -0,0 +1,16 @@
+
+/// ItemGroup:
+///   - an abstract base class for a group of items with a title.
+abstract class ItemGroup {
```
```suggestion
abstract class AbstractItemGroup {
```

Pertempto commented on lib/model/item_group.dart:
```diff
@@ -0,0 +1,16 @@
+
+  const ItemGroup();
+
+  Future<ListModelItemGroup> asListModelItemGroup();
```
```suggestion
  ItemGroup get originalGroup;
```

Pertempto replied:
I think the idea behind this was to be able to get the "original" item group from a filtered item group?

tycebrown replied:
Yes, the issue I was running into was due to the fact that in `ListWidget`, I store `groupsToBeDisplayed` as a `List` of `AbstractItemGroup` (so that I can display the groups initially gotten from the `listModel.groupsView()` or groups that have been filtered). When one of these groups is edited (say, the title is changed), the changes need to be put to isar. However, only the item groups stored in list models (and therefore the database) can be put, so I needed a way of getting an item group that can be put to the database from any abstract item group.

Do you want me to look into improving/completely-scrapping-and-redoing my design, at least in this area?

Pertempto replied:
In a normal database design, you would just store the ID of the original group, then use that ID to get the group that should be updated. You can redo your design, or just follow the renaming suggestions I gave you. I don't think your design is that bad, it just needs a few small improvements.

Pertempto commented on lib/model/item_group_search_results.dart:
```diff
@@ -0,0 +1,30 @@
+  Iterable<Item> itemsView() => _results;
+
+  @override
+  Future<ListModelItemGroup> asListModelItemGroup() =>
+      _group.asListModelItemGroup();
```
```suggestion
  get originalItemGroup() ->  _group;
```

Pertempto commented on lib/model/item_group_search_results.dart:
```diff
@@ -0,0 +1,30 @@
+  set title(String? newTitle) => _group.title = newTitle;
+
+  const ItemGroupSearchResults(
+      {required ItemGroup group, required Iterable<Item> results})
```
Leave this as `ItemGroup`. A filtered item group should only be based off of a full item group.

Pertempto reviewed (changes requested):
Here is my simple idea for how to make the whole Item Group class system make more sense.

* `AbstractItemGroup`
    * This defines the basic interface for an item group
* `ItemGroup extends AbstractItemGroup`
    * This is a normal item group, stored in a `ListModel`
* `FilteredItemGroup  extends AbstractItemGroup`
    * This is an item group that has had a filter applied to it.

First rename `ItemGroup` to `AbstractItemGroup`, otherwise you'll get some confusion between the new `ItemGroup` class and the old one.

Pertempto commented:
We need some left padding on the item group titles. It does not look good almost touching the left side! 😬

Pertempto commented:
When I open the app on Windows, I can create a new list. When I click the "back" button to go to the app home page, my new list has disappeared.

tycebrown commented:
> When I open the app on Windows, I can create a new list. When I click the "back" button to go to the app home page, my new list has disappeared.

Oh, that doesn't sound good. Will look into that...

Pertempto commented:
> Oh, that doesn't sound good. Will look into that...


Update: I found the isar files in `C:\Users\<username>\AppData\Roaming\lists.exe`, deleted them, and started the app again. Now everything works fine.

Pertempto commented:
What is the overall use case for categories in a list? What is an example of how a user would use them?

To me, it seems if they need a separate category, they should just create a whole new list. I might be missing some use case though.

tycebrown commented:
> What is the overall use case for categories in a list? What is an example of how a user would use them?
> 
> To me, it seems if they need a separate category, they should just create a whole new list. I might be missing some use case though.

I was partially inspired by [AnyList](https://apps.apple.com/us/app/anylist-grocery-shopping-list/id522167641). The use case I had in mind was in a grocery list, with, say, a "produce" category, a "dairy" category, a "meat" category, etc. 

Looking back, I see how frail my justification for categories are, as the user could just as well use multiple lists. Your thoughts?

tycebrown committed:
Lots of renaming, a few changes.

I decided to make EditItemGroupDialog take an ItemGroup, since we shouldn't need to edit an AbstractItemGroup.

I also changed ItemGroupWidget's onEdited function to take an ItemGroup (not Abstract). The Widget then passes the originalGroup to onEdited

Pertempto commented:
> I was partially inspired by [AnyList](https://apps.apple.com/us/app/anylist-grocery-shopping-list/id522167641). The use case I had in mind was in a grocery list, with, say, a "produce" category, a "dairy" category, a "meat" category, etc.
> 
> Looking back, I see how frail my justification for categories are, as the user could just as well use multiple lists. Your thoughts?

Ok! Your justification isn't frail. It just seems like this feature requires a lot of complicated changes. Is this feature worth those changes?  
If so, then great, let's keep going with this PR. I think we can make this work.  
If not, then we can close it. It would be disappointing to lose all the effort you put into this, but it wouldn't be wasted time. You gained valuable experience.

tycebrown commented:
> Ok! Your justification isn't frail. It just seems like this feature requires a lot of complicated changes. Is this feature worth those changes?
> If so, then great, let's keep going with this PR. I think we can make this work.
> If not, then we can close it. It would be disappointing to lose all the effort you put into this, but it wouldn't be wasted time. You gained valuable experience.

I'm not sure these features are worth the changes. I think we should close this, though I'm willing to be convinced otherwise.
